src/graph.rs

```diff
@@ -93,22 +96,53 @@ impl<'tree> Graph<'tree> {
 
     pub fn display_json<'a>(&'a self, ctx: &'a Context) -> () {
         struct JSONGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+        struct JSONNode<'a>(usize, &'a GraphNode, &'a Context);
+        struct JSONEdge<'a>(&'a u32, &'a Edge, &'a Context);
+
         impl<'a, 'tree> ser::Serialize for JSONGraph<'a, 'tree> {
             fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                 where S: serde::Serializer {
                 let graph = self.0;
                 let ctx = self.1;
-                let mut nodes = serializer.serialize_seq(Some(graph.graph_nodes.len()))?;
+                let mut seq = serializer.serialize_seq(Some(graph.graph_nodes.len()))?;
                 for (node_index, node) in graph.graph_nodes.iter().enumerate() {
-                    for (sink, edge) in &node.outgoing_edges {
-                        // )?;
-                    }
+                    seq.serialize_element(&JSONNode(node_index, node, ctx))?;
+                }
+                seq.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONNode<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where S: serde::Serializer {
+                let node_index = self.0;
+                let node = self.1;
+                let ctx = self.2;
+                // serializing as a map instead of a struct so we don't have to encode a struct name
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("index", &node_index)?;
+                if !node.outgoing_edges.is_empty() {
```

Comment:
I wonder if it is better always add the field, even if it is empty. It makes it easier for consumers, because there are fewer cases to handle.

Reply:
Yeah, I think uniformity of JSON output trumps the extremely minor expense of serializing an empty array.

Reply:
Good points. I was optimizing for human-readability, but that's really not our intended use case.